src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1040,49 +1038,73 @@ private enum AssignmentKind
             ParameterSymbol parameterOpt = null,
-            Conversion conversion = default,
             Location location = null)
         {
+            // Callers should apply any conversions before calling this method
+            // (see https://github.com/dotnet/roslyn/issues/39867).
+            if (targetType.HasType &&
+                !targetType.Type.Equals(valueType.Type, TypeCompareKind.AllIgnoreOptions))
+            {
+                return false;
+            }
+
             if (value == null ||
                 value.WasCompilerGenerated ||
                 !targetType.HasType ||
-                targetType.Type.IsValueType ||
-                targetType.CanBeAssignedNull ||
-                valueType.IsNotNull)
+                targetType.Type.IsValueType)
             {
                 return false;
             }
 
-            location ??= value.Syntax.GetLocation();
-            var unwrappedValue = SkipReferenceConversions(value);
-            if (unwrappedValue.IsSuppressed)
+            switch (targetType.NullableAnnotation)
             {
-                return false;
+                case NullableAnnotation.Oblivious:
+                case NullableAnnotation.Annotated:
+                    return false;
             }
 
-            if (RequiresSafetyWarningWhenNullIntroduced(targetType))
+            switch (valueType.State)
             {
-                HashSet<DiagnosticInfo> useSiteDiagnostics = null;
-                if (conversion.Kind == ConversionKind.UnsetConversionKind)
-                    conversion = this._conversions.ClassifyImplicitConversionFromType(valueType.Type, targetType.Type, ref useSiteDiagnostics);
-
-                if (conversion.IsImplicit && !conversion.IsDynamic)
-                {
-                    // For type parameters that cannot be annotated, the analysis must report those
-                    // places where null values first sneak in, like `default`, `null`, and `GetFirstOrDefault`,
-                    // as a safety diagnostic.  This is NOT one of those places.
+                case NullableFlowState.NotNull:
                     return false;
-                }
+                case NullableFlowState.MaybeNull:
+                    if (targetType.Type.IsTypeParameterDisallowingAnnotation())
+                    {
+                        return false;
+                    }
+                    break;
+            }
 
-                useLegacyWarnings = false;
+            location ??= value.Syntax.GetLocation();
+            var unwrappedValue = SkipReferenceConversions(value);
+            if (unwrappedValue.IsSuppressed)
+            {
+                return false;
             }
 
-            if (reportNullLiteralAssignmentIfNecessary(value, location, valueType.ToTypeWithAnnotations()))
+            if (useLegacyWarnings &&
+                valueType.Type?.TypeKind == TypeKind.TypeParameter &&
```

Comment:
>valueType.Type?.TypeKind == TypeKind.TypeParameter [](start = 16, length = 50)

This condition is insufficient.  We need to check that it is a type parameter that could not be annotated because of the absence of a suitable constraint.  See the implementation of `RequiresSafetyWarningWhenNullIntroduced` for the necessary condition.

Reply:
`valueType.State` will only be `MaybeDefault` if the type is `null` or a type parameter that cannot be annotated.

---
In reply to: [347681896](https://github.com/dotnet/roslyn/pull/39778#discussion_r347681896) [](ancestors = 347681896)